### PR TITLE
Webapp improvements

### PR DESCRIPTION
**Pull request by acheronw**
1. Filter by status functionality added to the webapp frontend.
2. More instructive error messages concerning the config.yaml.
3. Handling of ~/ directories in the config.yaml.
4. Added the requirements for the webapp to the setup.py.

### CODE REVIEW

File: app/crud.py [hunk 1]
```diff
@@ -44,38 +51,38 @@ def create_step(db: Session,
...
-        further_params += ' -P ' \
-                          + str(config['runtime_configurations']['processes'])
+    # These steps are marked by "no_p_param: True" in the config.yaml.
+    if not settings["scripts"][db_step.step_name].pop("no_p_param", False):
+        further_params += " -P " \
+                          + str(config["runtime_configurations"]["processes"])
 
     # Let's go over and process the parameters:
     for key, value in settings["scripts"][db_step.step_name].items():
-        if key == 'script_file':
```
Comment (DavidNemeskey) [inline app/crud.py:62]: I wouldn't mind double quotes, but we use single quotes everywhere else in the code...

File: app/logging.py [hunk 2]
```diff
@@ -1,11 +1,14 @@
 import logging
+from pathlib import Path
 
 from .config import config, get_logs_dir
 
 
 def configure_logging():
+    log_path = Path(get_logs_dir(config) / "server.log")
```
Comment (DavidNemeskey) [inline app/logging.py:8]: `get_logs_dir()` returns a `Path`, so you don't need the conversion here.

File: app/main.py [hunk 3]
```diff
@@ -6,7 +6,8 @@
 """
 
 
-from fastapi import BackgroundTasks, Depends, FastAPI, Form, Request, status
+from fastapi import BackgroundTasks, Depends, FastAPI, Form
+from fastapi import Query, Request, status
```
Comment (DavidNemeskey) [inline app/main.py:10]: You can also do it like this:
```Python
from fastapi import (
    ...
)
```

File: app/main.py [hunk 4]
```diff
@@ -69,10 +70,19 @@ def dev_seed():
...
+                status: str = Query(None),
+                db: Session = Depends(get_db)):
     """Controller for rendering the list of steps."""
-    steps = crud.get_steps(db)
-    context = {"request": request, "steps": steps}
+    if status:
+        filter = {'status': status}
+        steps = crud.get_steps(db, filters=filter)
+    else:
+        steps = crud.get_steps(db)
```
Comment (DavidNemeskey) [inline app/main.py:81]:
```Python
filter = {'status': status} if status else None
steps = crud.get_steps(db, filters=filter)
```
?

**Review by DavidNemeskey — CHANGES_REQUESTED** [review]

File: app/config.py [hunk 5]
```diff
@@ -17,14 +17,22 @@
...
+        raise FileNotFoundError('Config file is missing. Please make sure '
+                                'that you have created an '
+                                'app/config.yaml using the'
+                                'app/config_example.yaml as a template.')
 
     # If the working dir is not set properly we raise an error:
     working_dir_var = config["folders"]["working_dir"]
-    if not Path(working_dir_var).is_dir():
+    working_dir = Path(working_dir_var).expanduser()
+    if not Path(working_dir).is_dir():
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/config.py:32]: Ok.

File: app/config.py [hunk 6]
```diff
@@ -17,14 +17,22 @@
 
 def load_config_file(config_file) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/config.py:18]: Unforunately the values are not simply (Path | str), some of the fields contain dicts or lists, some recursively to the depth of 3. So I would keep the typehint of the values as "Any".

File: app/config.py [hunk 7]
```diff
@@ -17,14 +17,22 @@
...
 
     # If the working dir is not set properly we raise an error:
     working_dir_var = config["folders"]["working_dir"]
-    if not Path(working_dir_var).is_dir():
+    working_dir = Path(working_dir_var).expanduser()
+    if not Path(working_dir).is_dir():
         raise FileNotFoundError(
-            f"The working dir {working_dir_var} is missing.")
+            f"The working dir {working_dir} is missing. "
+            f"Please ensure that it is properly set in app/config.yaml")
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/config.py:35]: Ok.

File: app/config.py [hunk 8]
```diff
@@ -54,11 +62,11 @@ def load_config_with_defaults(config_file) -> Dict[str, Any]:
 
 
 def get_logs_dir(config) -> Path:
```
[the remark this answers is not in the document]
Comment (acheronw) [reply app/config.py:64]: Same problem as above, the values of the values (haha) are too varied, that's why the typehint is "Any".

File: app/crud.py (hunk 1 above)
Comment (acheronw) [reply app/crud.py:62]: Not anymore! :)

In the app/ I overwhelmingly used double quotes. So I unified it to use double quotes everywhere within the app/

File: app/logging.py (hunk 2 above)
Comment (acheronw) [reply app/logging.py:8]: Ok.

File: app/main.py (hunk 3 above)
Comment (acheronw) [reply app/main.py:10]: Good to know, thanks!

File: app/main.py (hunk 4 above)
Comment (acheronw) [reply app/main.py:81]: Ok.

File: setup.py [hunk 9]
```diff
@@ -107,6 +107,8 @@ def readme():
           # idzip for fast seek in compressed files
           # Uncommented until my PR is merged
           # 'idzip',
+          # Manager webapp framework:
+          'fastapi==0.101.1',
```
[the remark this answers is not in the document]
Comment (acheronw) [reply setup.py:111]: I changed it to  'fastapi>=0.101.1', if that was your point...

File: app/config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DavidNemeskey) [reply app/config.py:18]: I mean for `config_file` :smile:

File: app/config.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (DavidNemeskey) [reply app/config.py:64]: Same as above (and probably in other places as well)

File: app/crud.py (hunk 1 above)
Comment (DavidNemeskey) [reply app/crud.py:62]: OK, at least we'll have easy authorship attribution :laughing:

File: setup.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (DavidNemeskey) [reply setup.py:111]: Well, that's good; TBH I didn't really have a point, I was just flabbergasted at the minor version number.

File: app/config.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (acheronw) [reply app/config.py:18]: Ah, ok! :)

**Review by DavidNemeskey — APPROVED** [review]

**Pull request merged**